mypy `[arg-type]`-error when passing in type that can be converted to `RootModel`

### Initial Checks

- [x] I confirm that I'm using Pydantic V2

### Description

The below example code (named `foo.py`) passes when running `python foo.py` as the input `"1.0"`-string gets converted to `float`. However, running `mypy foo.py` (with or without `--strict`) gives the following error:
```bash
foo.py:11: error: Argument 1 to "RootModel" has incompatible type "str"; expected "float"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```
I would expect the `BaseModel`-fields and the `RootModel`-field to behave the same in this case. That is, if the value can be converted – either via a custom made validator or through the conversion table – then `mypy` should not raise an error. 

### Example Code

```Python
import pydantic


class Foo(pydantic.BaseModel):
    f: float


RFoo = pydantic.RootModel[float]

f = Foo(f="1.0")
rf = RFoo("1.0")

assert isinstance(f.f, float)
assert isinstance(rf.root, float)
assert f.f == rf.root
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.12.5
        pydantic-core version: 2.41.5
          pydantic-core build: profile=release pgo=false
               python version: 3.11.14 (main, Feb  3 2026, 22:51:56) [Clang 21.1.4 ]
                     platform: Linux-5.15.0-173-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.15.0 mypy-1.19.1
                       commit: unknown
```

--- a/pydantic/mypy.py
+++ b/pydantic/mypy.py
@@ -88,6 +88,7 @@ CREATE_MODEL_FULLNAME = 'pydantic.main.create_model'
 BASESETTINGS_FULLNAME = 'pydantic_settings.main.BaseSettings'
 ROOT_MODEL_FULLNAME = 'pydantic.root_model.RootModel'
 MODEL_METACLASS_FULLNAME = 'pydantic._internal._model_construction.ModelMetaclass'
+ROOT_MODEL_METACLASS_FULLNAME = 'pydantic.root_model._RootModelMetaclass'
 FIELD_FULLNAME = 'pydantic.fields.Field'
 DATACLASS_FULLNAME = 'pydantic.dataclasses.dataclass'
 MODEL_VALIDATOR_FULLNAME = 'pydantic.functional_validators.model_validator'
@@ -143,7 +144,7 @@ class PydanticPlugin(Plugin):
 
     def get_metaclass_hook(self, fullname: str) -> Callable[[ClassDefContext], None] | None:
         """Update Pydantic `ModelMetaclass` definition."""
-        if fullname == MODEL_METACLASS_FULLNAME:
+        if fullname in (MODEL_METACLASS_FULLNAME, ROOT_MODEL_METACLASS_FULLNAME):
             return self._pydantic_model_metaclass_marker_callback
         return None
 
@@ -916,7 +917,13 @@ class PydanticModelTransformer:
 
         The added `__init__` will be annotated with types vs. all `Any` depending on the plugin settings.
         """
-        if '__init__' in self._cls.info.names and not self._cls.info.names['__init__'].plugin_generated:
+        if (
+            '__init__' in self._cls.info.names
+            and not self._cls.info.names['__init__'].plugin_generated
+            # `RootModel` declares a typed `__init__` for the benefit of type checkers when the plugin
+            # isn't used, but the plugin-generated one should take priority over it:
+            and self._cls.fullname != ROOT_MODEL_FULLNAME
+        ):
             return  # Don't generate an __init__ if one already exists
 
         typed = self.plugin_config.init_typed

--- a/tests/mypy/modules/root_models.py
+++ b/tests/mypy/modules/root_models.py
@@ -22,6 +22,18 @@ pets1 = Pets1(['dog', 'cat'])
 pets2 = Pets2(['dog', 'cat'])
 pets3 = Pets3(['dog', 'cat'])
 
+RFloat = RootModel[float]
+
+
+class RFloatSub(RootModel[float]):
+    pass
+
+
+# When the plugin is used with `init_typed` unset, arbitrary input
+# should be accepted as it may be coerced to the `root` type:
+rfloat = RFloat('1.0')
+rfloat_sub = RFloatSub('1.0')
+
 
 class Pets4(RootModel[list[str]]):
     pets: list[str]

--- a/tests/mypy/outputs/mypy-default_ini/root_models.py
+++ b/tests/mypy/outputs/mypy-default_ini/root_models.py
@@ -23,6 +23,20 @@ pets1 = Pets1(['dog', 'cat'])
 pets2 = Pets2(['dog', 'cat'])
 pets3 = Pets3(['dog', 'cat'])
 
+RFloat = RootModel[float]
+
+
+class RFloatSub(RootModel[float]):
+    pass
+
+
+# When the plugin is used with `init_typed` unset, arbitrary input
+# should be accepted as it may be coerced to the `root` type:
+rfloat = RFloat('1.0')
+# MYPY: error: Argument 1 to "RootModel" has incompatible type "str"; expected "float"  [arg-type]
+rfloat_sub = RFloatSub('1.0')
+# MYPY: error: Argument 1 to "RFloatSub" has incompatible type "str"; expected "float"  [arg-type]
+
 
 class Pets4(RootModel[list[str]]):
     pets: list[str]

--- a/tests/mypy/outputs/mypy-plugin_ini/root_models.py
+++ b/tests/mypy/outputs/mypy-plugin_ini/root_models.py
@@ -23,6 +23,18 @@ pets1 = Pets1(['dog', 'cat'])
 pets2 = Pets2(['dog', 'cat'])
 pets3 = Pets3(['dog', 'cat'])
 
+RFloat = RootModel[float]
+
+
+class RFloatSub(RootModel[float]):
+    pass
+
+
+# When the plugin is used with `init_typed` unset, arbitrary input
+# should be accepted as it may be coerced to the `root` type:
+rfloat = RFloat('1.0')
+rfloat_sub = RFloatSub('1.0')
+
 
 class Pets4(RootModel[list[str]]):
     pets: list[str]

--- a/tests/mypy/outputs/pyproject-default_toml/root_models.py
+++ b/tests/mypy/outputs/pyproject-default_toml/root_models.py
@@ -23,6 +23,20 @@ pets1 = Pets1(['dog', 'cat'])
 pets2 = Pets2(['dog', 'cat'])
 pets3 = Pets3(['dog', 'cat'])
 
+RFloat = RootModel[float]
+
+
+class RFloatSub(RootModel[float]):
+    pass
+
+
+# When the plugin is used with `init_typed` unset, arbitrary input
+# should be accepted as it may be coerced to the `root` type:
+rfloat = RFloat('1.0')
+# MYPY: error: Argument 1 to "RootModel" has incompatible type "str"; expected "float"  [arg-type]
+rfloat_sub = RFloatSub('1.0')
+# MYPY: error: Argument 1 to "RFloatSub" has incompatible type "str"; expected "float"  [arg-type]
+
 
 class Pets4(RootModel[list[str]]):
     pets: list[str]

